### PR TITLE
bleergh

### PR DESCRIPTION
**Pull request by chrismccord**

### CODE REVIEW

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
+---
+layout: post
+title: "Optimizing Your Elixir and Phoenix projects with ETS"
+social: true
+author: Chris McCord
+twitter: "chris_mccord"
+github: chrismccord
+summary: "Learn how to optimize your elixir applications with a fast in-memory store"
+published: true
+tags: elixir, phoenix, ets
```
Comment (bcardarella) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]:
```diff
+ engineering
```

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+This just barely scratched the surface of what features ETS provides. We'll only be using a fraction of its capabilities, but just remember where it really shines in fast reads and writes to key-value storage, with the ability to efficiently match on most erlang terms stored within the table (excluding maps). In our examples, we'll only be storing simply key-values with basic lookups, but you should consult the docs to explore the breadth of provided features.
```
Comment (alexgaribay) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: but just remember where it really shines ~~in~~ **is** fast reads

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+use it. Projects across the community like `Phoenix.PubSub`,
+`Phoenix.Presence`, and Elixir's own `Registry` take advantage of ETS, along with many internal Erlang and Elixir modules.
+Simply put, ETS is an in-memory store for Elixir and Erlang terms with
+fast access. Critically, it allows access to state outside of a
+process and message passing. Before we talk about what ETS can do, let's begin with a basic primer by firing up iex:
+
+First, we'll create an ETS table with `:ets.new/2`:
+
+
+```
```
Comment (bcardarella) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: the code fences can take language. Can you add `elixir` to each?

Comment (bcardarella) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: actually, looks like this is meant to be IEX... I *think* the language still makes sense

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+This just barely scratched the surface of what features ETS provides. We'll only be using a fraction of its capabilities, but just remember where it really shines in fast reads and writes to key-value storage, with the ability to efficiently match on most erlang terms stored within the table (excluding maps). In our examples, we'll only be storing simply key-values with basic lookups, but you should consult the docs to explore the breadth of provided features.
+
+
+## Optimizing `GenServer` access with an ETS table
+
+Optimizing code is a rewarding experince, but doubly so when you don't have to change your public interface. Let's see a common way ETS is used to optimize data access for state wrapped in a `GenServer`.
+
+Imagine we're writing a rate limiter `GenServer` which is a process in our app that counts user requests and allows us to deny access once a user exceeds their allotted requests per minute. We know right away that we'll need to store the request count state for our users somewhere, as well as a process that periodically sweeps the state once per minute. A naive first-pass with a plain-old `GenServer` might look something like this:
+
+```
```
Comment (bcardarella) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: definitely code fenced language here

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+[]
+iex> RateLimiter.log("user1")
+:ok
+```
+
+It works just as before. We also use `:ets.tab2list/1` to spy on the data in the table. We can see our users requests are tracked properly, and the table is swept as expected.
+
+That's all there to it. Our public interface remained unchanged and we vastly improve the performance of our mission-critical feature. Not bad!
+
+This just scratched the surface on what's possible with ETS. If you're curious about using ETS for a dependency-free in-memory cache, check out Saša Jurić's excellent [ConCache](https://github.com/sasa1977/con_cache) library.
```
Comment (bcardarella) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: is ConnCache popular? I trust what Saša writes, but I've just always used `:ets` directly

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md (hunk 4 above)
Comment (chrismccord) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: done!

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md (hunk 5 above)
Comment (chrismccord) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: yes. It if you are using ets as a basic in memory cache with expiration, con_cache is what you want

Comment (chrismccord) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: The expiration part is key and tricky to get right without races, so definitely recommended. If you aren't worried about expiry, then not as strong argument to use it vs ets directly

File: source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md (hunk 3 above)
Comment (chrismccord) [inline source/posts/2017-05-19-optimizing-elixir-and-phoenix-with-ets.md]: fencing with iex does not provide any highlighting

**Pull request merged**